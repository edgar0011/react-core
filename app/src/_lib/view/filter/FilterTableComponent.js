import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'

export default class FilterTableComponent extends PureComponent {
  static instanceNum = 0
  static propTypes = {
    name: PropTypes.string,
    labels: PropTypes.object.isRequired,
    tableData: PropTypes.shape({
      columns: PropTypes.array.isRequired,
      data: PropTypes.arrayOf(PropTypes.shape({
        id: PropTypes.oneOfType([PropTypes.number, PropTypes.string]).isRequired })),
      options: PropTypes.shape({
        tableId: PropTypes.string.isRequired,
        selectable: PropTypes.bool,
        allSelectable: PropTypes.bool,
        noHeader: PropTypes.bool,
        light: PropTypes.bool,
        onSortClick: PropTypes.func,
        onRowClick: PropTypes.func,
        onHeaderClick: PropTypes.func,
        onCellClick: PropTypes.func,
        onSelectRow: PropTypes.func,
        onSelectAllRows: PropTypes.func
      }).isRequired,
      renderPagination: PropTypes.func,
      selection: PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.number, PropTypes.string])),
      sorting: PropTypes.shape({ columnTitle: PropTypes.string, direction: PropTypes.string })
    }),
    formData: PropTypes.shape({
      fields: PropTypes.arrayOf(PropTypes.shape({
        name: PropTypes.string.isRequired,
        label: PropTypes.string.isRequired
      }).isRequired),
      errors: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.shape({
        code: PropTypes.string.isRequired,
        message: PropTypes.string
      })), PropTypes.any]),
      onChangeHandler: PropTypes.func,
      onBlurHandler: PropTypes.func,
      onFocusHandler: PropTypes.func,
      onSubmitHandler: PropTypes.func.isRequired,
      onResetHandler: PropTypes.func.isRequired,
      enableToggleMode: PropTypes.bool
    }),

    additionalRender: PropTypes.shape({
      top: PropTypes.func,
      middle: PropTypes.func,
      bottom: PropTypes.func,
      foot: PropTypes.func,
      renderPagination: PropTypes.func
    }),
    loading: PropTypes.bool,
    paginationProps: PropTypes.shape({
      initialPage: PropTypes.number,
      pageCount: PropTypes.number,
      pageSize: PropTypes.number,
      totalResults: PropTypes.number,
      onChangePage: PropTypes.func.isRequired
    }),
    translations: PropTypes.object,
    showPagination: PropTypes.bool,
    showExport: PropTypes.bool,
    exportLoading: PropTypes.bool,
    onUnmountHandler: PropTypes.func,
    components: PropTypes.shape({
      Filter: PropTypes.any,
      TablePagination: PropTypes.any,
      Table: PropTypes.any,
      Export: PropTypes.any,
      noDataComponent: PropTypes.any,
      loadingDataComponent: PropTypes.any,
      Loader: PropTypes.any,
      LoadingData: PropTypes.any
    })
  }

  static defaultProps = {
    translations: {}
  }

  constructor(props) {
    super(props)
    FilterTableComponent.instanceNum += 1
    this.instanceName = `filterTable-${FilterTableComponent.instanceNum}`
  }

  componentWillUnmount() {
    if (this.props.onUnmountHandler) {
      this.props.onUnmountHandler()
    }
  }

  renderPaginationMethodPropsWrapper = (tableProps) => {
    const tablePaginationProps = {
      paginationProps: this.props.paginationProps,
      showExport: this.props.showExport,
      exportLoading: this.props.exportLoading,
      translations: this.props.translations,
      ...tableProps
    }

    const renderPaginationMethod = this.props.additionalRender && this.props.additionalRender.renderPagination ?
      this.props.additionalRender.renderPagination : null

    if (renderPaginationMethod) {
      return renderPaginationMethod(tablePaginationProps)
    }
    return <this.props.components.TablePagination {...tablePaginationProps} />
  }

  render() {
    const { name, formData, tableData, labels, loading, additionalRender, showPagination,
      components: {
        noDataComponent, loadingDataComponent, Filter, Table, Export, Loader, LoadingData
      }, showExport, translations } = this.props

    const NoDataCompo = noDataComponent || <h1><LoadingData description={labels.noData || 'No results found...'} /></h1>
    const LoadingDataCompo = loadingDataComponent || <h1><LoadingData description={labels.loadingData} /></h1>

    if (showExport) {
      // TODO if rightTabContent is already set, wrap around
      tableData.rightTabContent = <Export {...this.props} />
    }
    tableData.renderPagination = showPagination ? this.renderPaginationMethodPropsWrapper : null

    return (
      <div id={name || this.instanceName}>
        {additionalRender && additionalRender.top && additionalRender.top({ ...this.props })}
        {formData &&
        <Filter
          formName={formData.formName || `${name || this.instanceName}-form`}
          {...formData}
          translations={translations}
        />}
        {additionalRender && additionalRender.middle && additionalRender.middle({ ...this.props })}
        {loading &&
        <Loader text={labels.loading} />}
        {(tableData.data && tableData.data.length > 0) &&
        <Table
          tableData={tableData}
        />}
        {(!loading && (!tableData.data || !tableData.data.length)) && NoDataCompo}
        {(loading && LoadingDataCompo && !tableData.data) && LoadingDataCompo}
        {additionalRender && additionalRender.bottom && additionalRender.bottom({ ...this.props })}
        {additionalRender && additionalRender.foot && additionalRender.foot({ ...this.props })}
      </div>
    )
  }
}
