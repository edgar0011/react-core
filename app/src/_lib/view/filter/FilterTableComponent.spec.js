import React from 'react'
import FilterTableComponent from './FilterTableComponent'

describe('FilterTableComponent', () => {
  it('should render FilterTableComponent', () => {
    const props = {
      components: {},
      labels: {},
      tableData: {},
      formData: {}
    }
    global.shallowSnapshot(<FilterTableComponent {...props} />)
  })
})
