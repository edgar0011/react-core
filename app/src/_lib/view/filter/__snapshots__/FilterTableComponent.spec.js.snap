// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterTableComponent should render FilterTableComponent 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <FilterTableComponent
    components={Object {}}
    formData={Object {}}
    labels={Object {}}
    tableData={
      Object {
        "renderPagination": null,
      }
    }
    translations={Object {}}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        undefined,
        <UNDEFINED
          formName="filterTable-1-form"
          translations={Object {}}
        />,
        undefined,
        undefined,
        undefined,
        <h1>
          <UNDEFINED
            description="No results found..."
          />
        </h1>,
        undefined,
        undefined,
        undefined,
      ],
      "id": "filterTable-1",
    },
    "ref": null,
    "rendered": Array [
      undefined,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "formName": "filterTable-1-form",
          "translations": Object {},
        },
        "ref": null,
        "rendered": null,
        "type": undefined,
      },
      undefined,
      undefined,
      undefined,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <UNDEFINED
            description="No results found..."
          />,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "description": "No results found...",
          },
          "ref": null,
          "rendered": null,
          "type": undefined,
        },
        "type": "h1",
      },
      undefined,
      undefined,
      undefined,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          undefined,
          <UNDEFINED
            formName="filterTable-1-form"
            translations={Object {}}
          />,
          undefined,
          undefined,
          undefined,
          <h1>
            <UNDEFINED
              description="No results found..."
            />
          </h1>,
          undefined,
          undefined,
          undefined,
        ],
        "id": "filterTable-1",
      },
      "ref": null,
      "rendered": Array [
        undefined,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "formName": "filterTable-1-form",
            "translations": Object {},
          },
          "ref": null,
          "rendered": null,
          "type": undefined,
        },
        undefined,
        undefined,
        undefined,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <UNDEFINED
              description="No results found..."
            />,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "description": "No results found...",
            },
            "ref": null,
            "rendered": null,
            "type": undefined,
          },
          "type": "h1",
        },
        undefined,
        undefined,
        undefined,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
